File: Python/1082-read_number_in_chinese.py


digit_to_hanzi = ['ling', 'yi', 'er', 'san', 'si',
                  'wu', 'liu', 'qi', 'ba', 'jiu']
pos_to_thousands = ['', 'Shi', 'Bai', 'Qian', 'Wan',
                    'Shi', 'Bai', 'Qian', 'Yi']


def reading_number(n):
    if n == 0:
        return 'ling'
    n = str(n)
    if n[0] == '-':
        sign = 'Fu '
        n = n[1:]
    else:
        sign = ''
    res = []
    zeros = 0
    for i, digit in enumerate(n, start=1):
        pos = len(n) - i  # counting from right
        if digit != '0':
            if zeros > 0:
                res.append('ling')
            res.append(digit_to_hanzi[int(digit)])
            if pos > 0:  # no need to output '个'位
                res.append(pos_to_thousands[pos])
            zeros = 0
        else:
            zeros += 1
            if pos == 4 and zeros < 4:  # handling numbers like '100001'
                res.append('Wan')
    return sign + ' '.join(res)


if __name__ == '__main__':
    n = int(input())
    print(reading_number(n))

File: C++/1082-read_number_in_chinese.cc
#include <iostream>
#include <string>
#include <array>


std::array<std::string, 10> digit_to_hanzi = {
    "ling", "yi", "er", "san", "si", "wu", "liu", "qi", "ba", "jiu"};
std::array<std::string, 9> pos_to_thounsands = {
    "", "Shi", "Bai", "Qian", "Wan", "Shi", "Bai", "Qian", "Yi"};


std::string reading_number(int n) {
    if (n == 0) return "ling";
    std::string res;
    if (n < 0) {
        res += "Fu";
        n *= -1;
    }
    std::string n_str = std::to_string(n);
    int zeros = 0;
    for (std::size_t i = 0; i < n_str.length(); ++i) {
        std::size_t pos = n_str.length() - i - 1;  // counting from right
        int digit = n_str[i] - '0';
        if (digit != 0) {
            if (zeros > 0) res += " ling";
            if (res.size() > 0) {
                res += " " + digit_to_hanzi[digit];
            } else {
                res += digit_to_hanzi[digit];
            }
            // no need to output '个'位
            if (pos > 0) res += " " + pos_to_thounsands[pos];
            zeros = 0;
        } else {
            zeros += 1;
            // handling numbers like '100001'
            if (pos == 4 && zeros < 4) res += " Wan";
        }
    }
    return res;
}


int main() {

    int n;
    std::cin >> n;
    std::cout << reading_number(n) << '\n';

    return 0;
}
